src/Hosting/Hosting/src/StaticWebAssets/StaticWebAssetsFileProvider.cs

```diff
@@ -40,20 +44,32 @@ public StaticWebAssetsFileProvider(string pathPrefix, string contentRoot)
         /// <inheritdoc />
         public IDirectoryContents GetDirectoryContents(string subpath)
         {
-            if (!StartsWithBasePath(subpath, out var physicalPath))
+            var modifiedSub = NormalizePath(subpath);
+
+            if (StartsWithBasePath(modifiedSub, out var physicalPath))
             {
-                return NotFoundDirectoryContents.Singleton;
+                return InnerProvider.GetDirectoryContents(physicalPath.Value);
+            }
+            else if (string.Equals(subpath, string.Empty))
+            {
+                return new StaticWebAssetsDirectoryRoot(BasePath);
+            }
+            else if (BasePath.StartsWithSegments(modifiedSub, FilePathComparison, out PathString remaining))
+            {
+                return new StaticWebAssetsDirectoryRoot(remaining);
             }
             else
             {
-                return InnerProvider.GetDirectoryContents(physicalPath.Value);
+                return NotFoundDirectoryContents.Singleton;
             }
```

Comment:
You could simplify this block as below. Also, why do we need to special case the empty string. Does `BasePath.StartsWithSegments(modifiedSub, FilePathComparison, out PathString remaining)` not handle it? Finally, when you are dealing with a `PathString` instance you need to pass in the `Value` property, otherwise it will http encode the value and things with spaces in it will fail.

```chsarp
            if (string.Equals(subpath, string.Empty))
            {
                return new StaticWebAssetsDirectoryRoot(BasePath);
            }

            var modifiedSub = NormalizePath(subpath);
            if (StartsWithBasePath(modifiedSub, out var physicalPath))
            {
                return InnerProvider.GetDirectoryContents(physicalPath.Value);
            }
            
            if (BasePath.StartsWithSegments(modifiedSub, FilePathComparison, out PathString remaining))
            {
                return new StaticWebAssetsDirectoryRoot(remaining.Value);
            }
            
            return NotFoundDirectoryContents.Singleton;
```

Reply:
`BasePath.StartsWithSegments` does not handle the empty string case. I'm glad you brought that up though, it also doesn't handle the case of `/` properly, so that will need a case too (also adding tests).

I'm fuzzier on the `PathString` part, because the constructor for `StaticWebAssetsDirectoryRoot` takes a `PathString` (which it [only uses the `Value` property](https://github.com/aspnet/AspNetCore/pull/14803/files#diff-11cbff5b8d8860e80be8908027ec3c07R104) off of). If we do `remaining.Value` that string will just get [implicitly converted](https://github.com/aspnet/AspNetCore/blob/master/src/Http/Http.Abstractions/src/PathString.cs#L460) back to a `PathString`. It's definitely worth adding an explanatory comment to the constructor about the importance of using that for future generations though, and by keeping that constructor parameter a `PathString` we only have to get this right in one spot.